src/game/room.c

```diff
@@ -590,19 +590,42 @@ int16_t Room_GetWaterHeight(int32_t x, int32_t y, int32_t z, int16_t room_num)
 
 void Room_AlterFloorHeight(ITEM_INFO *item, int32_t height)
 {
-    int16_t room_num = item->room_number;
-    FLOOR_INFO *floor =
-        Room_GetFloor(item->pos.x, item->pos.y, item->pos.z, &room_num);
-    FLOOR_INFO *ceiling = Room_GetFloor(
-        item->pos.x, item->pos.y + height - WALL_L, item->pos.z, &room_num);
-
-    if (floor->floor == NO_HEIGHT / 256) {
-        floor->floor = ceiling->ceiling + height / 256;
-    } else {
-        floor->floor += height / 256;
-        if (floor->floor == ceiling->ceiling && floor->sky_room == NO_ROOM) {
+    if (!height) {
+        return;
+    }
+
+    int16_t data;
+    FLOOR_INFO *floor;
+    ROOM_INFO *r = &g_RoomInfo[item->room_number];
+
+    do {
+        int32_t z_floor = (item->pos.z - r->z) >> WALL_SHIFT;
+        int32_t x_floor = (item->pos.x - r->x) >> WALL_SHIFT;
+
+        if (z_floor <= 0) {
+            z_floor = 0;
+            CLAMP(x_floor, 1, r->y_size - 2);
+        } else if (z_floor >= r->x_size - 1) {
+            z_floor = r->x_size - 1;
+            CLAMP(x_floor, 1, r->y_size - 2);
+        } else {
+            CLAMP(x_floor, 0, r->y_size - 1);
+        }
```

Comment:
Why do we clamp the floor to 1… size-2 rather than 0…size-1?

Reply:
Check out [Room_GetFloor](https://github.com/LostArtefacts/TR1X/blob/2078ede62e4c88141f202fabb87915dacac4dfcb/src/game/room.c#L233). This is extremely similar to that except it excludes the Y axis traversal.

The slight difference is I used `x_floor` and `z_floor` because the floor traversal is across those axes. And now that also matches TRView. I think the OG code used x/y as in x/y plane, but it was slightly confusing because there is also the 3D world's Y axis.